clang/include/clang/Tooling/DesignatedInitializers.h

```diff
@@ -0,0 +1,43 @@
+//===--- DesignatedInitializers.h -------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides utilities for designated initializers.
+///
+//===----------------------------------------------------------------------===//
+
+#include "clang/AST/Expr.h"
+#include "clang/Basic/SourceLocation.h"
+#include "llvm/ADT/DenseMap.h"
+
+namespace clang::tooling {
+
+/// Get designators describing the elements of a (syntactic) init list.
+///
+/// Given for example the type
+///
+///     struct S { int i, j; };
+///
+/// and the definition
+///
+///     S s{1, 2};
+///
+/// calling `getDesignators` for the initializer list expression `{1, 2}`
+/// would produce the map `{loc(1): ".i", loc(2): ".j"}`.
+///
+/// It does not produce designators for any explicitly-written nested lists,
+/// e.g. `{1, .j=2}` would only return `{loc(1): ".i"}`.
+///
+/// It also considers structs with fields of record types like
+/// `struct T { S s; };`. In this case, there would be designators of the
+/// form
+/// `.s.i` and `.s.j` in the returned map.
+llvm::DenseMap<clang::SourceLocation, std::string>
+getDesignators(const clang::InitListExpr *Syn);
```

Comment:
I'd suggest renaming this function to `getUnwrittenDesignators` because that is what it does. It does not return existing designators (I know the name comes from the clangd implementation).

Reply:
Do you have a comment on the location? Is it okay to have it in `clang/Tooling`?

Reply:
I don't think it is a good location to have this file here because of what the other files contain. One possibility would be to add this into `clang-tools-extra/clang-tidy/utils` because clangd links the clang-tidy (clangTidy) in its cmake. Another reviewer should confirm this though.

Reply:
That's also an option, but i think that clangd link to clang-tidy is configurable.

Reply:
As I understand it, the base of clang-tidy is always linked into Clangd, but linking in all the checks can be disabled. `clangTidyUtils` is another library anyway and would need to be linked into Clangd separately if the files are moved to `clang-tools-extra/clang-tidy/utils`.

I have done so in https://github.com/llvm/llvm-project/pull/80541/commits/df24916e9dd25b3f8b7dc03dcb782e689ea1af37. Please let me know if this looks correct to you.

Reply:
@PiotrZSL: One last confirmation please. 🙂